# Análise de dados estruturados

In [2]:
import pandas as pd

![panda](https://media.giphy.com/media/HDR31jsQUPqQo/giphy.gif)

## Criar um dataframe a partir de um dicionário

In [1]:
dict = {"country": ["Brazil", "Russia", "India", "China", "South Africa"],
       "capital": ["Brasilia", "Moscow", "New Dehli", "Beijing", "Pretoria"],
       "area": [8.516, 17.10, 3.286, 9.597, 1.221],
       "population": [200.4, 143.5, 1252, 1357, 52.98] }

In [3]:
# Transformar o dicionário em um dataframe
brics = pd.DataFrame(dict)

In [4]:
# Ver os primeiros registros desse dataframe
brics.head()

area    capital       country  population
0   8.516   Brasilia        Brazil      200.40
1  17.100     Moscow        Russia      143.50
2   3.286  New Dehli         India     1252.00
3   9.597    Beijing         China     1357.00
4   1.221   Pretoria  South Africa       52.98

## Importar um csv com o pandas

Vamos utilizar os dados que o Kaggle lançou no ano de 2017 sobre Cientistas de Dados e Data Science. São 5 datasets diferentes:

 - **schema.csv**: a CSV file with survey schema. This schema includes the questions that correspond to each column name in both the multipleChoiceResponses.csv and freeformResponses.csv.
 - **multipleChoiceResponses.csv**: Respondents' answers to multiple choice and ranking questions. These are non-randomized and thus a single row does correspond to all of a single user's answers. -freeformResponses.csv: Respondents' freeform answers to Kaggle's survey questions. These responses are randomized within a column, so that reading across a single row does not give a single user's answers.
 - **conversionRates.csv**: Currency conversion rates (to USD) as accessed from the R package "quantmod" on September 14, 2017
 - **RespondentTypeREADME.txt**: This is a schema for decoding the responses in the "Asked" column of the schema.csv file.

In [16]:
# Carregue o dataset multipleChoiceResponses com o pandas 
multiple_choice = pd.read_csv('kaggle-survey-2017/multipleChoiceResponses.csv')

In [17]:
# Veja as primeiras linhas do dataset
multiple_choice.head()

GenderSelect        Country   Age  \
0  Non-binary, genderqueer, or gender non-conforming            NaN   NaN   
1                                             Female  United States  30.0   
2                                               Male         Canada  28.0   
3                                               Male  United States  56.0   
4                                               Male         Taiwan  38.0   

                                    EmploymentStatus StudentStatus  \
0                                 Employed full-time           NaN   
1                 Not employed, but looking for work           NaN   
2                 Not employed, but looking for work           NaN   
3  Independent contractor, freelancer, or self-em...           NaN   
4                                 Employed full-time           NaN   

  LearningDataScience CodeWriter CareerSwitcher  \
0                 NaN        Yes            NaN   
1                 NaN        NaN            NaN   
2                 NaN        NaN            NaN   
3                 NaN        Yes            NaN   
4                 NaN        Yes            NaN   

              CurrentJobTitleSelect TitleFit              ...                \
0             DBA/Database Engineer     Fine              ...                 
1                               NaN      NaN              ...                 
2                               NaN      NaN              ...                 
3  Operations Research Practitioner   Poorly              ...                 
4                Computer Scientist     Fine              ...                 

  JobFactorExperienceLevel JobFactorDepartment  JobFactorTitle  \
0                      NaN                 NaN             NaN   
1                      NaN                 NaN             NaN   
2           Very Important      Very Important  Very Important   
3                      NaN                 NaN             NaN   
4                      NaN                 NaN             NaN   

  JobFactorCompanyFunding JobFactorImpact JobFactorRemote JobFactorIndustry  \
0                     NaN             NaN             NaN               NaN   
1                     NaN             NaN             NaN               NaN   
2          Very Important  Very Important  Very Important    Very Important   
3                     NaN             NaN             NaN               NaN   
4                     NaN             NaN             NaN               NaN   

  JobFactorLeaderReputation JobFactorDiversity JobFactorPublishingOpportunity  
0                       NaN                NaN                            NaN  
1        Somewhat important                NaN                            NaN  
2            Very Important     Very Important                 Very Important  
3                       NaN                NaN                            NaN  
4                       NaN                NaN                            NaN  

[5 rows x 228 columns]

In [23]:
#Veja a quantidade de linhas e de colunas do dataset
multiple_choice.shape

(16716, 228)

Existem 228 colunas!!!
![panda](https://media.giphy.com/media/14aUO0Mf7dWDXW/giphy.gif)

Vamos ver do que se tratam essas colunas. Como são MUITAS colunas precisamos alterar a configuração padrão do pandas para visualização de linhas e colunas

In [32]:
pd.set_option('max_rows', 1000)
pd.set_option('max_columns', 1000)

In [34]:
multiple_choice.head()

GenderSelect        Country   Age  \
0  Non-binary, genderqueer, or gender non-conforming            NaN   NaN   
1                                             Female  United States  30.0   
2                                               Male         Canada  28.0   
3                                               Male  United States  56.0   
4                                               Male         Taiwan  38.0   

                                    EmploymentStatus StudentStatus  \
0                                 Employed full-time           NaN   
1                 Not employed, but looking for work           NaN   
2                 Not employed, but looking for work           NaN   
3  Independent contractor, freelancer, or self-em...           NaN   
4                                 Employed full-time           NaN   

  LearningDataScience CodeWriter CareerSwitcher  \
0                 NaN        Yes            NaN   
1                 NaN        NaN            NaN   
2                 NaN        NaN            NaN   
3                 NaN        Yes            NaN   
4                 NaN        Yes            NaN   

              CurrentJobTitleSelect TitleFit  \
0             DBA/Database Engineer     Fine   
1                               NaN      NaN   
2                               NaN      NaN   
3  Operations Research Practitioner   Poorly   
4                Computer Scientist     Fine   

                                 CurrentEmployerType MLToolNextYearSelect  \
0  Employed by a company that doesn't perform adv...             SAS Base   
1                                                NaN               Python   
2                                                NaN  Amazon Web services   
3                                      Self-employed           TensorFlow   
4  Employed by a company that doesn't perform adv...           TensorFlow   

  MLMethodNextYearSelect LanguageRecommendationSelect  \
0         Random Forests                           F#   
1         Random Forests                       Python   
2          Deep learning                            R   
3            Neural Nets                       Python   
4            Text Mining                       Python   

                                PublicDatasetsSelect  \
0  Dataset aggregator/platform (i.e. Socrata/Kagg...   
1  Dataset aggregator/platform (i.e. Socrata/Kagg...   
2  Dataset aggregator/platform (i.e. Socrata/Kagg...   
3          I collect my own data (e.g. web-scraping)   
4                                             GitHub   

                              LearningPlatformSelect  \
0  College/University,Conferences,Podcasts,Trade ...   
1                                             Kaggle   
2  Arxiv,College/University,Kaggle,Online courses...   
3  Blogs,College/University,Conferences,Friends n...   
4                  Arxiv,Conferences,Kaggle,Textbook   

  LearningPlatformUsefulnessArxiv LearningPlatformUsefulnessBlogs  \
0                             NaN                             NaN   
1                             NaN                             NaN   
2                     Very useful                             NaN   
3                             NaN                     Very useful   
4                     Very useful                             NaN   

  LearningPlatformUsefulnessCollege LearningPlatformUsefulnessCompany  \
0                               NaN                               NaN   
1                               NaN                               NaN   
2                   Somewhat useful                               NaN   
3                       Very useful                               NaN   
4                               NaN                               NaN   

  LearningPlatformUsefulnessConferences LearningPlatformUsefulnessFriends  \
0                           Very useful                               NaN   
1                                   NaN                               Na

Podemos ver só o nome das colunas também utilizando o `columns`. Para ficar mais fácil de visualizar, ao invés de retornar o array, podemos transformar esse dado em uma Series.

In [35]:
# Use o columns no dataframe e coloque-o em uma Series para facilitar a visualização
pd.Series(multiple_choice.columns)

0                                     GenderSelect
1                                          Country
2                                              Age
3                                 EmploymentStatus
4                                    StudentStatus
5                              LearningDataScience
6                                       CodeWriter
7                                   CareerSwitcher
8                            CurrentJobTitleSelect
9                                         TitleFit
10                             CurrentEmployerType
11                            MLToolNextYearSelect
12                          MLMethodNextYearSelect
13                    LanguageRecommendationSelect
14                            PublicDatasetsSelect
15                          LearningPlatformSelect
16                 LearningPlatformUsefulnessArxiv
17                 LearningPlatformUsefulnessBlogs
18               LearningPlatformUsefulnessCollege
19               LearningPlatfo

O que eu devo fazer se eu quiser ver apenas coluna `JobFactorSalary`?

In [36]:
multiple_choice['JobFactorSalary']

0                       NaN
1                       NaN
2            Very Important
3                       NaN
4                       NaN
5                       NaN
6                       NaN
7            Very Important
8                       NaN
9                       NaN
10       Somewhat important
11                      NaN
12       Somewhat important
13                      NaN
14                      NaN
15                      NaN
16                      NaN
17                      NaN
18       Somewhat important
19            Not important
20       Somewhat important
21                      NaN
22                      NaN
23                      NaN
24                      NaN
25                      NaN
26                      NaN
27                      NaN
28                      NaN
29           Very Important
30                      NaN
31                      NaN
32                      NaN
33                      NaN
34                      NaN
35           Very Im

Como eu faço se eu só quiser ver os 10 primeiros registros?

In [41]:
multiple_choice['JobFactorSalary'][:10]

0               NaN
1               NaN
2    Very Important
3               NaN
4               NaN
5               NaN
6               NaN
7    Very Important
8               NaN
9               NaN
Name: JobFactorSalary, dtype: object

E se eu quiser ver 2 colunas ao mesmo tempo? (E apenas essas 2 colunas)

In [40]:
multiple_choice[['JobFactorSalary', 'JobFactorLearning']][:10]

JobFactorSalary JobFactorLearning
0             NaN               NaN
1             NaN               NaN
2  Very Important    Very Important
3             NaN               NaN
4             NaN               NaN
5             NaN               NaN
6             NaN               NaN
7  Very Important    Very Important
8             NaN               NaN
9             NaN               NaN

O quanto que as pessoas dessa pesquisa estão satisfeitas com o trabalhos? Conseguimos saber isso usando só o pandas?

In [43]:
multiple_choice['JobSatisfaction'].value_counts()

7                          1448
8                          1427
6                           765
9                           677
5                           627
10 - Highly Satisfied       589
3                           358
4                           354
1 - Highly Dissatisfied     167
I prefer not to share       148
2                           117
Name: JobSatisfaction, dtype: int64

Percebemos com esse comando que as pessoas até que estão bastante satisfeitas.

Agora vamos olhar só as pessoas que estão Super Satisfeitas (Highly Satisfied) com o seu trabalho. Como que eu posso fazer isso?

In [50]:
# Filtre só quem está com o JobSatisfaction de 10. Guarde isso em uma variável pq é bastante dado
highly_satisfied = multiple_choice[multiple_choice['JobSatisfaction'] == '10 - Highly Satisfied']

In [53]:
# veja o tamanho do dataset. Ele bateu com a quantidade de pessoas que estão altamente satisfeitas?
highly_satisfied.shape

(589, 228)

In [51]:
# Veja os primeiros 3 registros das pessoas altamentes satisfeitas
highly_satisfied[:3]

GenderSelect        Country   Age  \
3          Male  United States  56.0   
60         Male         Canada  34.0   
77         Male         Israel  29.0   

                                     EmploymentStatus StudentStatus  \
3   Independent contractor, freelancer, or self-em...           NaN   
60  Independent contractor, freelancer, or self-em...           NaN   
77  Independent contractor, freelancer, or self-em...           NaN   

   LearningDataScience CodeWriter CareerSwitcher  \
3                  NaN        Yes            NaN   
60                 NaN        Yes            NaN   
77                 NaN        Yes            NaN   

               CurrentJobTitleSelect   TitleFit CurrentEmployerType  \
3   Operations Research Practitioner     Poorly       Self-employed   
60                             Other       Fine       Self-employed   
77                    Data Scientist  Perfectly       Self-employed   

   MLToolNextYearSelect MLMethodNextYearSelect LanguageRecommendationSelect  \
3            TensorFlow            Neural Nets                       Python   
60                  NaN                    NaN                       Python   
77                  NaN          Deep learning                       Python   

                                 PublicDatasetsSelect  \
3           I collect my own data (e.g. web-scraping)   
60                                                NaN   
77  Dataset aggregator/platform (i.e. Socrata/Kagg...   

                               LearningPlatformSelect  \
3   Blogs,College/University,Conferences,Friends n...   
60  Arxiv,Blogs,Kaggle,Online courses,Personal Pro...   
77  Arxiv,Kaggle,Official documentation,Online cou...   

   LearningPlatformUsefulnessArxiv LearningPlatformUsefulnessBlogs  \
3                              NaN                     Very useful   
60                             NaN                             NaN   
77                 Somewhat useful                             NaN   

   LearningPlatformUsefulnessCollege LearningPlatformUsefulnessCompany  \
3                        Very useful                               NaN   
60                               NaN                               NaN   
77                               NaN                               NaN   

   LearningPlatformUsefulnessConferences LearningPlatformUsefulnessFriends  \
3                            Very useful                       Very useful   
60                                   NaN                               NaN   
77                                   NaN                               NaN   

   LearningPlatformUsefulnessKaggle LearningPlatformUsefulnessNewsletters  \
3                               NaN                                   NaN   
60                              NaN                                   NaN   
77                      Very useful                                   NaN   

   LearningPlatformUsefulnessCommunities  \
3                                    NaN   
60                                   NaN   
77                                   NaN   

   LearningPlatformUsefulnessDocumentation LearningPlatformUsefulnessCourses  \
3                              Very useful                       Very useful   
60                                     NaN                               NaN   
77                             Very useful                       Very useful   

   LearningPlatformUsefulnessProjects LearningPlatformUsefulnessPodcasts  \
3                         Very useful                                NaN   
60                                NaN                                NaN   
77                                NaN                                NaN   

   LearningPlatformUsefulnessSO LearningPlatformUsefulnessTextbook  \
3                           NaN                                NaN   
60                          NaN                                NaN   
77                  Very useful                                NaN   

   L

E se eu quiser ver as pessoas altamente satisfeitas e que trabalham com python?

In [122]:
highly_satisfied = multiple_choice['JobSatisfaction'] == '10 - Highly Satisfied'
pythonist = multiple_choice['LanguageRecommendationSelect'] == 'Python'
highly_satisfied_and_pythonist = multiple_choice[highly_satisfied & learn_importance]

In [121]:
highly_satisfied_and_pythonist.shape

(319, 228)

E se tentassemos com a idade? Ver só que está abaixo de 30 anos

In [118]:
highly_satisfied = multiple_choice['JobSatisfaction'] == '10 - Highly Satisfied'
age = multiple_choice['Age'] < 30.0
highly_satisfied_and_age = multiple_choice[highly_satisfied & age]

In [119]:
highly_satisfied_and_age.shape

(171, 228)

Quais são as linguagens que a galera altamente satisfeita trabalha?

In [125]:
multiple_choice[highly_satisfied]['LanguageRecommendationSelect'].value_counts()

Python      319
R           151
SQL          28
C/C++/C#     20
Matlab       11
Other         9
Scala         9
Java          7
Julia         5
SAS           4
Stata         3
Haskell       2
Name: LanguageRecommendationSelect, dtype: int64

In [126]:
highly_satisfied_languages = multiple_choice[highly_satisfied]['LanguageRecommendationSelect'].value_counts()
language_counts = multiple_choice['LanguageRecommendationSelect'].value_counts()

In [127]:
highly_satisfied_languages / language_counts

C/C++/C#    0.065147
F#               NaN
Haskell     0.117647
Java        0.050725
Julia       0.166667
Matlab      0.046218
Other       0.105882
Python      0.045959
R           0.057132
SAS         0.045455
SQL         0.072727
Scala       0.095745
Stata       0.107143
Name: LanguageRecommendationSelect, dtype: float64

Precisa ordenar

In [128]:
## adicionar a coluna da currency

Objetivo da aula: Aprender a estruturar e manipular dados a partir de diferentes fontes,
utilizando o Pandas e fazer análise desses dados utilizando os métodos e funções da
biblioteca

Outcomes esperados:

● Discutir as aplicações do uso da biblioteca Pandas em projetos reais de Ciência de
Dados, trazendo exemplos de empresas de tecnologia

● Aprofundar os métodos e funções da biblioteca Pandas utilizados para manipulação
e análise de dados estruturados

● Praticar a importação, leitura, organização e manipulação de dados de diferentes
fontes para utilizar o Pandas para análise desse dataset

Perguntas-chave:

● Como preparar os dados estruturados para manipular utilizando o Pandas?

● Quais são os métodos e funções mais utilizados do Pandas em projetos de Ciência
de Dados?

● Em que estágios de um projeto a biblioteca pode ajudar o Cientista de Dados a
manipular dados estruturados?